### PR TITLE
feat: email signup, login and account verification

### PR DESCRIPTION
**Pull request by Karlmusingo**
#### What does this PR do?
- signup with email
- login with username and password
- account verification using email

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,6 +45,7 @@
     "supertest": "^4.0.2",
     "swagger-ui-express": "^4.0.4",
     "unique-slug": "^2.0.1",
+    "unirest": "^0.6.0",
```
Comment (oesukam) [inline package.json:47]: @Karlmusingo Why this yet we have `supertest`?

File: src/__tests__/routes/organization.test.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+describe('Organization auth', () => {
+  describe('Signup with email', () => {
+    test('should sign up the user', async () => {
+      const resp = await request(app)
+        .post(`${urlPrefix}/auth/signup`)
+        .send(organization);
+      expect(resp.status).toBe(201);
+      expect(resp.body).toHaveProperty('token');
+    });
```
Comment (oesukam) [inline src/__tests__/routes/organization.test.js:21]: @Karlmusingo New line required between test cases

File: src/__tests__/routes/organization.test.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      const badOrg = organization;
+      badOrg.username = 'new_username';
+      const resp = await request(app)
+        .post(`${urlPrefix}/auth/signup`)
+        .send(badOrg);
+      expect(resp.status).toBe(400);
+      expect(resp.body.message).toBe('Email already exist');
+    });
+  });
+  describe('Email verification', () => {
```
Comment (oesukam) [inline src/__tests__/routes/organization.test.js:42]: New line required to separate functions

File: src/controllers/AuthController.js [hunk 4]
```diff
@@ -37,9 +39,7 @@ class AuthController {
         .exec();
 
     const username = displayName.replace(/\s+/g, '') + id.substr(0, 5);
-    const token = jwt.sign({ email, username }, process.env.SECRET, {
-      expiresIn: '2d',
-    });
+    const token = encrypt.generateToken({ email, username });
```
Comment (oesukam) [inline src/controllers/AuthController.js]: @Karlmusingo Why did you change `jwt` to `encrypt`?

File: src/controllers/AuthController.js [hunk 5]
```diff
@@ -105,6 +105,146 @@ class AuthController {
...
+   * @memberof Auth
+   */
+  static async signup(req, res) {
+    const hashedPassword = encrypt.hashPassword(req.body.password);
+    const { companyName, username, email } = req.body;
+
+    const ret = field =>
+      res.status(400).json({ message: `${field} already exist` });
+
+    const existUsername = await User.findOne({
```
Comment (oesukam) [inline src/controllers/AuthController.js]: @Karlmusingo Should be checked from a middleware. Try to extract the finding a user to another file

File: src/controllers/AuthController.js [hunk 6]
```diff
@@ -105,6 +105,146 @@ class AuthController {
...
+
+    const getUser = id =>
+      Organization.findById(id)
+        .populate('user')
+        .exec();
+
+    const result = await getUser(organization.id);
+
+    const token = encrypt.generateToken({ email, username });
+    if (process.env.NODE_ENV !== 'test') {
```
Comment (oesukam) [inline src/controllers/AuthController.js]: @Karlmusingo Why this?

File: src/controllers/AuthController.js [hunk 7]
```diff
@@ -105,6 +105,146 @@ class AuthController {
...
+   * @static
+   * @param {*} req
+   * @param {*} res
+   * @returns {object} res
+   * @memberof Auth
+   */
+  static async verification(req, res) {
+    try {
+      const { token } = req.params;
+      const jwtPayload = jwt.verify(token, process.env.SECRET);
```
Comment (oesukam) [inline src/controllers/AuthController.js]: @Karlmusingo Why not have a middleware for this?

File: src/helpers/encrypt.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  /**
+   * @param  {Object} data
+   * @returns  {string}} {expiresIn
+   */
+  static generateToken(data) {
+    const token = jwt.sign(
+      data,
+      process.env.SECRET,
```
Comment (oesukam) [inline src/helpers/encrypt.js]: Should be destructed from the top of the file

File: src/helpers/sendMail.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * property
+ *
+ * @param {string} email
+ * @param {string} name
+ * @param {string} token
+ * @param {string} body
+ * @returns {object} {sent,error}
+ */
+const sendToken = async (email, name, token, body = bodyDefault) => {
+  const link = `${process.env.DOMAIN}/api/v1/auth/verification/${token}`;
```
Comment (oesukam) [inline src/helpers/sendMail.js]: DOMAIN should be destructed

**Review by oesukam — CHANGES_REQUESTED** [review]

File: src/controllers/AuthController.js (hunk 4 above)
Comment (Karlmusingo) [reply src/controllers/AuthController.js]: I am using it at different places, so I decided to put it in the helpers folder.

File: package.json (hunk 1 above)
Comment (Karlmusingo) [reply package.json:47]: I used it for requesting the sendgrid API

File: src/controllers/AuthController.js (hunk 6 above)
Comment (Karlmusingo) [reply src/controllers/AuthController.js]: To not try to send emails when running test

File: src/controllers/AuthController.js (hunk 4 above)
Comment (oesukam) [reply src/controllers/AuthController.js]: OK, but you should only sign using the id without providing the username and email

File: src/controllers/AuthController.js (hunk 6 above)
Comment (oesukam) [reply src/controllers/AuthController.js]: @Karlmusingo Test should not be handled in your logic, but it should be mocked instead

File: src/__tests__/routes/organization.test.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        .send(badOrg);
+      expect(resp.status).toBe(409);
+      expect(resp.body.message).toBe('Email already exist');
+    });
+  });
+
+  describe('Email verification', () => {
+    const { email, username } = organization;
+    const token = encrypt.generateToken({ email, username });
+    test('should log the user in when the account is not verified', async () => {
```
Comment (gracelungu) [inline src/__tests__/routes/organization.test.js]: @Karlmusingo, This test forbids an unverified account to log in and does not log in the user, as specified in the test description.

File: src/controllers/AuthController.js [hunk 11]
```diff
@@ -105,6 +104,116 @@ class AuthController {
...
+        .populate('user')
+        .exec();
+
+    const result = await getUser(organization.id);
+
+    const token = encrypt.generateToken({ email, username });
+    if (process.env.NODE_ENV !== 'test') {
+      await sendMail(email, companyName, token);
+    }
+    return res.status(201).json({
```
Comment (gracelungu) [inline src/controllers/AuthController.js]: @Karlmusingo, the token should also be returned in the body, in some cases, this may be needed for those who are requesting from mediums that do not provide the status of the response, for example, someone fetching a GET request directly from the browser.

**Review by gracelungu — CHANGES_REQUESTED** [review]

**Comment by gracelungu** [comment]
[Users should be able to login](https://trello.com/c/1pmJshF6/1-users-should-be-able-to-login)

**Comment by gracelungu** [comment]
[Users should be able to login](https://trello.com/c/1pmJshF6/1-users-should-be-able-to-login)
